cpp/comms/std/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,408 @@
+  memcpy(id->internal, uniqueId, size);
+}
+
+/**
+ * @brief Returns a NCCL unique ID as a character array. PyTorch
+ * uses this same approach, so that it can be more easily
+ * converted to a native Python string by Cython and further
+ * serialized to be sent across process & node boundaries.
+ *
+ * @returns the generated NCCL unique ID for establishing a
+ * new clique.
+ */
+void get_unique_id(char *uid, int size) {
+  ncclUniqueId id;
+  ncclGetUniqueId(&id);
+
+  memcpy(uid, id.internal, size);
+}
+
+#ifdef WITH_UCX
+cumlStdCommunicator_impl::cumlStdCommunicator_impl(
+  ncclComm_t comm, ucp_worker_h ucp_worker, std::shared_ptr<ucp_ep_h *> eps,
+  int size, int rank)
+  : _nccl_comm(comm),
+    _ucp_worker(ucp_worker),
+    _ucp_eps(eps),
+    _size(size),
+    _rank(rank),
+    _next_request_id(0) {
+  initialize();
+}
+#endif
+
+cumlStdCommunicator_impl::cumlStdCommunicator_impl(ncclComm_t comm, int size,
+                                                   int rank)
+  : _nccl_comm(comm), _size(size), _rank(rank) {
+  initialize();
+}
+
+void cumlStdCommunicator_impl::initialize() {
+  cudaStreamCreate(&stream);
+
+  CUDA_CHECK(cudaMalloc((void **)&sendbuff, sizeof(int)));
+  CUDA_CHECK(cudaMalloc((void **)&recvbuff, sizeof(int)));
+}
+
+
+
+cumlStdCommunicator_impl::~cumlStdCommunicator_impl() {
+
+  cudaStreamDestroy(stream);
+
+  CUDA_CHECK_NO_THROW(cudaFree(sendbuff));
+  CUDA_CHECK_NO_THROW(cudaFree(recvbuff));
+}
+
+int cumlStdCommunicator_impl::getSize() const { return _size; }
+
+int cumlStdCommunicator_impl::getRank() const { return _rank; }
+
+std::unique_ptr<MLCommon::cumlCommunicator_iface>
+cumlStdCommunicator_impl::commSplit(int color, int key) const {
+  // Not supported by NCCL
+  ASSERT(
+    false,
+    "ERROR: commSplit called but not supported in this comms implementation.");
+}
+
+void cumlStdCommunicator_impl::barrier() const {
+  CUDA_CHECK(cudaMemsetAsync(sendbuff, 0, sizeof(int), stream));
+  CUDA_CHECK(cudaMemsetAsync(recvbuff, 0, sizeof(int), stream));
+
+  allreduce(sendbuff, recvbuff, 1, MLCommon::cumlCommunicator::INT,
+            MLCommon::cumlCommunicator::SUM, stream);
+
+  cudaStreamSynchronize(stream);
+}
+
+void cumlStdCommunicator_impl::isend(const void *buf, int size, int dest,
+                                     int tag, request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  request_t req_id;
+  if (_free_requests.empty())
+    req_id = _next_request_id++;
+  else {
+    auto it = _free_requests.begin();
+    req_id = *it;
+    _free_requests.erase(it);
+  }
+
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
+
+  ucp_ep_h *ep_arr = *_ucp_eps;
+  ucp_ep_h ep_ptr = ep_arr[dest];
+
+  struct ucx_context *ucp_request = ucp_isend(ep_ptr, buf, size, tag);
+
+  _requests_in_flight.insert(std::make_pair(req_id, ucp_request));
+  *request = req_id;
+
+  ucs_status_t flush_status = flush_ep(_ucp_worker, ep_ptr);
```

Comment:
Should `isend` be flushing the endpoint? Won't that make this function a `send`?

Reply:
Ah, okay. This makes sense. Another thing I took directly from the `hello_world.c` class.

Reply:
Actually, I was using a blocking variant of a non-blocking call here. I've replaced this to use ucp_ep_flush_nb with a callback. Are you saying I probably don't need to flush at all, even if it's nonblocking?

Reply:
Verified that I don't seem to need the flushing so removing for now.